File: backend/app/agent/tools.py
"""What the agent can actually do, and the guard rails on each.

The tools themselves are not new. The workspace already knew how to read a
file, propose a write and apply it; the terminal already knew how to run a
command and how to tell a command you typed from one a model produced. What
was missing was any way for a model to reach them and, more importantly, to
find out what happened afterwards.

Every tool here returns a plain result the model is shown. That is the whole
point: an agent is not a model that can act, it is a model that can act and
then read the consequence. A file write that reports "written" and a file
write that reports "no such directory" have to look different to the model,
or it cannot correct itself and the work stops at the first mistake.

Nothing here reaches outside the folder the user opened. The workspace
resolves every path against that root and refuses anything that climbs out of
it, so a wrong path fails as a message rather than as a file somewhere else on
the disk.

Which folder that is has to be decided by the caller, not assumed. The desktop
app has one open folder and a single Workspace to match. An editor extension
has whatever project the person is looking at, which is frequently not the
same one - and an agent that took the app's folder would have written the
editor's changes into somebody else's project. So every tool is given the
workspace to act in, and the caller says which.
"""

from __future__ import annotations

import json
import logging
import os
import subprocess
from typing import Any, Callable, Dict, List, Optional

logger = logging.getLogger("agent.tools")

#: How much of a file or a command's output the model is shown. Long enough to
#: work with, short enough that one large file cannot fill the context and
#: push the actual instruction out of it.
MAX_OUTPUT_CHARS = 12000


class ToolError(RuntimeError):
    """Something the model asked for could not be done, with the reason."""


def _clip(text: str, limit: int = MAX_OUTPUT_CHARS) -> str:
    text = text or ""
    if len(text) <= limit:
        return text
    return text[:limit] + "\n… (%d more characters not shown)" % (len(text) - limit)


# ---------------------------------------------------------------------------
# The tools
# ---------------------------------------------------------------------------

def workspace_for(root: str = ""):
    """The folder this run works in.

    With a root, a Workspace of its own opened there - which is how the editor
    extension gets the project the person is actually looking at rather than
    whatever the desktop app happens to have open. Without one, the app's
    single open folder, unchanged.

    Opening enforces the same refusals either way: not a drive root, not a home
    directory, and every path resolved back inside the root afterwards.
    """
    if (root or "").strip():
        from app.workspace.core import Workspace, WorkspaceError

        own = Workspace()
        try:
            own.open(root)
        except WorkspaceError as exc:
            raise ToolError(str(exc)) from exc
        return own

    from app.workspace.core import workspace

    if not workspace.describe().get("open"):
        raise ToolError(
            "No folder is open, so there is nothing to work in. Open one first "
            "- the interface has 'Open a folder' above the message box."
        )
    return workspace


def list_files(workspace, path: str = "") -> str:
    """Everything in the workspace, or under one directory of it."""
    tree = workspace.tree()
    entries = tree.get("entries") or tree.get("tree") or []
    if path:
        wanted = path.strip("/\\")
        entries = [e for e in entries
                   if str(e.get("path", "")).replace("\\", "/").startswith(wanted)]
    if not entries:
        return "Nothing found under %r." % (path or ".")
    lines = ["%s%s" % (e.get("path"), "/" if e.get("type") == "dir" else "")
             for e in entries]
    return _clip("\n".join(lines))


def read_file(workspace, path: str) -> str:
    """The contents of one file, with line numbers so edits can refer to them."""
    result = workspace.read(path)
    # The workspace calls this "text". Reading a key that is not there returned
    # the empty string, so every file looked empty and edit_file could never
    # find anything to replace. Watching a run is what caught it: the model
    # read two files, was told both were empty, and went round the problem by
    # printing them with a shell command.
    text = result["text"]
    numbered = "\n".join("%5d  %s" % (i, line)
                         for i, line in enumerate(text.splitlines(), start=1))
    return _clip(numbered) or "(the file is empty)"


def write_file(workspace, path: str, content: str) -> str:
    """Create or replace a file.

    Goes through the workspace's proposal mechanism and is then applied, so the
    change is recorded and reversible rather than written straight over
    somebody's work with no trace.
    """
    proposal = workspace.propose_write(path, content, summary="written by the agent")
    change_id = proposal.get("id") or proposal.get("change_id")
    if not change_id:
        raise ToolError("The write could not be prepared: %s" % json.dumps(proposal)[:200])
    workspace.apply(change_id)
    lines = content.count("\n") + 1
    return "Wrote %s (%d lines)." % (path, lines)


def edit_file(workspace, path: str, find: str, replace: str) -> str:
    """Replace an exact piece of text in a file.

    Exact rather than fuzzy, and it refuses when the text appears more than
    once. A model that is slightly wrong about the surrounding lines should be
    told so, not have its guess applied to whichever match came first.
    """
    current = workspace.read(path)["text"]
    occurrences = current.count(find)
    if occurrences == 0:
        return ("That exact text is not in %s. Read the file again and copy the "
                "lines you mean, including their indentation." % path)
    if occurrences > 1:
        return ("That text appears %d times in %s, so it is not clear which one "
                "you mean. Include more surrounding lines to make it unique."
                % (occurrences, path))
    return write_file(workspace, path, current.replace(find, replace, 1))


def search(workspace, query: str, path: str = "") -> str:
    """Find which files contain a piece of text."""
    root = workspace.resolve(path or ".")
    found: List[str] = []
    for base, dirs, files in os.walk(root):
        dirs[:] = [d for d in dirs
                   if d not in {".git", "node_modules", "__pycache__", ".venv", "dist"}]
        for name in files:
            full = os.path.join(base, name)
            try:
                with open(full, "r", encoding="utf-8", errors="ignore") as handle:
                    for number, line in enumerate(handle, start=1):
                        if query in line:
                            found.append("%s:%d: %s"
                                         % (os.path.relpath(full, root).replace("\\", "/"),
                                            number, line.strip()[:160]))
                            break
            except OSError:
                continue
            if len(found) >= 80:
                break
    return _clip("\n".join(found)) if found else "No file contains %r." % query


def run_command(workspace, command: str) -> str:
    """Run a shell command in the workspace and return what it printed.

    The output is the point. A command whose result the model never sees is a
    command it cannot learn from - it would write a test, run it, and never
    find out whether it passed.
    """
    root = str(workspace.root)
    try:
        finished = subprocess.run(
            command, shell=True, cwd=root, capture_output=True,
            text=True, timeout=300,
        )
    except subprocess.TimeoutExpired:
        return "The command was still running after five minutes and was stopped."
    except OSError as exc:
        raise ToolError("The command could not start: %s" % exc) from exc

    output = (finished.stdout or "") + (finished.stderr or "")
    return _clip("exit code %d\n%s" % (finished.returncode, output.strip() or "(no output)"))


def git(workspace, subcommand: str) -> str:
    """Run one git command in the workspace.

    Separate from run_command so that version control is something the agent
    can be given or refused on its own, and so "commit and push" is a thing it
    knows how to do rather than a shell command it has to guess at.
    """
    return run_command(workspace, "git " + subcommand)


# ---------------------------------------------------------------------------
# What the model is told it can do
# ---------------------------------------------------------------------------

#: Name -> (function, required argument names, one line for the model).
TOOLS: Dict[str, tuple] = {
    "list_files":  (list_files,  ["path"],
                    "List files in the workspace. path is optional."),
    "read_file":   (read_file,   ["path"],
                    "Read a file. Returns it with line numbers."),
    "write_file":  (write_file,  ["path", "content"],
                    "Create or completely replace a file."),
    "edit_file":   (edit_file,   ["path", "find", "replace"],
                    "Replace an exact piece of text in a file. The text must "
                    "appear exactly once."),
    "search":      (search,      ["query"],
                    "Find which files contain a piece of text."),
    "run_command": (run_command, ["command"],
                    "Run a shell command in the workspace and read its output."),
    "git":         (git,         ["subcommand"],
                    "Run a git command, for example: status, add -A, "
                    "commit -m \"...\", push."),
}

#: Tools that change something. Listed so the caller can decide which of them
#: need a person to agree first - writing a file and reading one are not the
#: same kind of act.
MUTATING = {"write_file", "edit_file", "run_command", "git"}


def describe_tools() -> str:
    """The tool list as the model is shown it."""
    lines = []
    for name, (_, args, description) in TOOLS.items():
        lines.append("- %s(%s): %s" % (name, ", ".join(args), description))
    return "\n".join(lines)


def execute(name: str, arguments: Dict[str, Any], workspace) -> str:
    """Run one tool in the given workspace, and return what the model sees."""
    entry = TOOLS.get(name)
    if entry is None:
        return ("There is no tool called %r. The ones that exist are: %s."
                % (name, ", ".join(TOOLS)))

    function, required, _ = entry
    missing = [a for a in required if a not in arguments and a != "path"]
    if missing:
        return "%s needs %s." % (name, " and ".join(missing))

    accepted = function.__code__.co_varnames[:function.__code__.co_argcount]
    try:
        return function(workspace, **{k: v for k, v in arguments.items()
                                      if k in accepted and k != "workspace"})
    except ToolError as exc:
        return str(exc)
    except Exception as exc:  # noqa: BLE001 - the model is shown the failure
        logger.warning("tool %s failed: %s", name, exc, exc_info=True)
        return "%s failed: %s" % (name, str(exc)[:300])

File: vscode-extension/src/agent/tools.ts
/**
 * What the agent can do, inside the folder you have open and nowhere else.
 *
 * These are the same seven tools the desktop app has, written again here so
 * that the extension needs nothing else installed. That duplication is a real
 * cost and worth naming: the app's copy is in backend/app/agent/tools.py, and
 * a change to one is a change to both. It is the price of the extension
 * working on a machine that has only VS Code on it, which is the point.
 *
 * The boundary is the folder. Every path is resolved first and then checked to
 * still be inside the root - that order matters, because `root/link/../../etc`
 * looks perfectly fine as text and is not fine at all once resolved.
 */

import { unified } from './diff';
import * as browser from './browser';
import { exec } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

/** Long enough to work with, short enough that one file cannot fill the context. */
const MAX_OUTPUT = 12000;

/** Reading a file into a prompt has to stop somewhere. */
const MAX_READ_BYTES = 400_000;

/** Large, generated, and never what somebody means by "my project". */
const SKIP = new Set([
    '.git', '.hg', '.svn', 'node_modules', '__pycache__', '.venv', 'venv',
    'env', 'dist', 'build', '.next', '.nuxt', 'target', '.gradle', '.idea',
    '.mypy_cache', '.pytest_cache', '.ruff_cache', '.tox', 'site-packages',
    '.terraform', 'vendor', 'Pods', '.cache', 'out',
]);

export class ToolError extends Error {}

function clip(text: string, limit = MAX_OUTPUT): string {
    if (text.length <= limit) {
        return text;
    }
    return `${text.slice(0, limit)}\n… (${text.length - limit} more characters not shown)`;
}

/**
 * A path inside the open folder - or anywhere, when reach says so.
 *
 * The check resolves symlinks first, so a link pointing out of the project is
 * refused for where it lands rather than allowed for where it sits. When reach
 * is set to Anywhere the check is skipped, which is what that setting means;
 * the path is still resolved, so what runs is a real absolute path and not
 * whatever relative string the model happened to write.
 */
let confined = true;

/** Set once per run, from the reach dial. */
export function confineToFolder(on: boolean): void {
    confined = on;
}

function resolveInside(root: string, relative: string): string {
    const candidate = path.resolve(root, relative || '.');
    if (!confined) {
        return candidate;
    }

    let real = candidate;
    let realRoot = root;
    try {
        realRoot = fs.realpathSync(root);
        real = fs.existsSync(candidate)
            ? fs.realpathSync(candidate)
            : path.join(fs.realpathSync(path.dirname(candidate)), path.basename(candidate));
    } catch {
        // A parent that does not exist yet is reported by the caller as a
        // missing directory, which is more useful than a resolution failure.
    }

    if (real !== realRoot && !real.startsWith(realRoot + path.sep)) {
        throw new ToolError(
            `${relative} is outside the open folder. Only files under ${root} can be read or changed.`,
        );
    }
    return real;
}

function walk(root: string, under: string): string[] {
    const found: string[] = [];
    const visit = (dir: string) => {
        let entries: fs.Dirent[];
        try {
            entries = fs.readdirSync(dir, { withFileTypes: true });
        } catch {
            return;
        }
        for (const entry of entries) {
            if (entry.isDirectory() && SKIP.has(entry.name)) {
                continue;
            }
            const full = path.join(dir, entry.name);
            const shown = path.relative(root, full).split(path.sep).join('/');
            if (entry.isDirectory()) {
                found.push(`${shown}/`);
                if (found.length < 2000) {
                    visit(full);
                }
            } else {
                found.push(shown);
            }
        }
    };
    visit(under);
    return found;
}

export const TOOLS: Record<string, { args: string[]; description: string; changes: boolean }> = {
    list_files: { args: ['path'], description: 'List files in the project. path is optional.', changes: false },
    read_file: { args: ['path'], description: 'Read a file. Returns it with line numbers.', changes: false },
    write_file: { args: ['path', 'content'], description: 'Create or completely replace a file.', changes: true },
    edit_file: {
        args: ['path', 'find', 'replace'],
        description: 'Replace an exact piece of text in a file. The text must appear exactly once.',
        changes: true,
    },
    search: { args: ['query'], description: 'Find which files contain a piece of text.', changes: false },
    run_command: { args: ['command'], description: 'Run a shell command in the project and read its output.', changes: true },
    git: { args: ['subcommand'], description: 'Run a git command, for example: status, add -A, commit -m "...", push.', changes: true },
    /* Eyes.
     *
     * Without these the agent could build a page and never learn whether
     * it works: it wrote the code, said it was done, and the person
     * opened it themselves and found it broken. The loop that the agents
     * this one is compared to actually run is open it, look, fix, look
     * again - and there was nothing here to look with.
     *
     * Reading only. Clicking and typing are a larger surface and can be
     * added once reading has earned its place. */
    open_browser: {
        args: ['url'],
        description: 'Open a page in a real browser window and watch it. Use this after starting a dev server, and whenever the task is about something a person would look at. Example url: localhost:3000',
        changes: false,
    },
    browser_check: {
        args: [],
        description: 'Report what the open page says: uncaught errors, console errors and warnings, requests that failed, and the text it rendered. This is how you find out whether your change worked.',
        changes: false,
    },
    browser_click: {
        args: ['text'],
        description: 'Click something by the words on it - the label a person would read, for example "Save" or "Sign in". Not a CSS selector. If more than one thing matches it clicks nothing and tells you what it saw.',
        changes: true,
    },
    browser_type: {
        args: ['into', 'text', 'enter'],
        description: 'Type into a field, found by its label or placeholder. Replaces whatever is already in it. Set enter to yes to press Enter afterwards.',
        changes: true,
    },
    browser_reload: {
        args: [],
        description: 'Reload the open page and report it again. Use this after editing a file, to see whether the problem is gone.',
        changes: false,
    },
    /* The step list, published by the model and kept on screen.
     *
     * A long run was a scrolling column of tool calls: every step visible, the
     * shape of the work not. Watching it, you could not tell whether the agent
     * was on its second step of five or its fifth of five, or whether it had
     * quietly dropped the half of the task you cared about.
     *
     * Antigravity calls this a Task List and writes it before any code;
     * Claude Code keeps a todo list through the run. Same idea both times, and
     * the value is the same: the plan is stated where it can be checked
     * against what actually happens.
     *
     * It changes nothing on disk, so it is never gated by a mode. */
    todo: {
        args: ['items'],
        description: 'Publish or update your step list. items is one step per line, each '
            + 'starting with [ ] for not done, [~] for in progress, or [x] for done. '
            + 'Send the whole list every time, not just what changed. Use it at the '
            + 'start of anything with more than two steps, and again as each one is done.',
        changes: false,
    },
};

/** One published step. */
export interface TodoItem {
    state: 'todo' | 'doing' | 'done';
    text: string;
}

/**
 * Read the model's step list.
 *
 * Deliberately forgiving about the marker: models write `[ ]`, `[]`, `- [ ]`,
 * `1. [x]`, and a bullet with no box at all. A list that renders as nothing
 * because a dash was missing would be worse than no list, so anything with
 * text on the line becomes an item, and only the marker decides its state.
 */
export function parseTodo(text: string): TodoItem[] {
    const items: TodoItem[] = [];
    for (const line of String(text || '').split('\n')) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        const marked = /^(?:[-*]\s*|\d+[.)]\s*)?\[([ x~*-]?)\]\s*(.+)$/i.exec(trimmed);
        if (marked) {
            const mark = marked[1].toLowerCase();
            items.push({
                state: mark === 'x' ? 'done' : (mark === '~' || mark === '*') ? 'doing' : 'todo',
                text: marked[2].trim(),
            });
            continue;
        }
        const bare = /^(?:[-*]\s+|\d+[.)]\s+)(.+)$/.exec(trimmed);
        if (bare) {
            items.push({ state: 'todo', text: bare[1].trim() });
        }
    }
    return items;
}

export function describeTools(): string {
    return Object.entries(TOOLS)
        .map(([name, tool]) => `- ${name}(${tool.args.join(', ')}): ${tool.description}`)
        .join('\n');
}

function listFiles(root: string, args: Record<string, string>): string {
    const under = args.path ? resolveInside(root, args.path) : root;
    if (!fs.existsSync(under)) {
        return `There is nothing at ${args.path}.`;
    }
    const entries = walk(root, under);
    return entries.length ? clip(entries.join('\n')) : `Nothing found under ${args.path || '.'}.`;
}

function readFile(root: string, args: Record<string, string>): string {
    const target = resolveInside(root, args.path);
    if (!fs.existsSync(target)) {
        return `${args.path} does not exist.`;
    }
    const stat = fs.statSync(target);
    if (stat.isDirectory()) {
        return `${args.path} is a folder, not a file.`;
    }
    if (stat.size > MAX_READ_BYTES) {
        return `${args.path} is ${(stat.size / 1e6).toFixed(1)} MB, past the ${(MAX_READ_BYTES / 1e6).toFixed(1)} MB limit for reading.`;
    }
    const text = fs.readFileSync(target, 'utf8');
    if (!text) {
        return '(the file is empty)';
    }
    const numbered = text
        .split('\n')
        .map((line, index) => `${String(index + 1).padStart(5)}  ${line}`)
        .join('\n');
    return clip(numbered);
}

/* Every change, with what was there before it.
 *
 * Two reasons, and the second is the one that matters. A diff needs the old
 * text to compare against, and an agent that edits a file has destroyed the
 * only copy of it unless something kept one. Without this, "put that back"
 * had no answer except the version control the project may not have.
 *
 * Bounded, because a long run rewriting a large file would otherwise hold
 * every version of it in memory at once. The oldest is dropped first, which
 * is also the one least likely to be wanted.
 */
interface Recorded {
    path: string;
    before: string;
    after: string;
    existed: boolean;
    at: number;
}

const MAX_REMEMBERED = 40;
const history: Recorded[] = [];

function record(entry: Recorded): void {
    history.push(entry);
    if (history.length > MAX_REMEMBERED) history.shift();
}

/** The changes made so far, newest last. */
export function changes(): ReadonlyArray<Recorded> {
    return history;
}

/**
 * Put the most recent change back, and say what was undone.
 *
 * It refuses when the file has moved on since - somebody may have edited it
 * by hand, and silently throwing that away to restore an older version is
 * worse than the change being undone.
 */
export function undoLast(root: string): string {
    const last = history.pop();
    if (!last) { return "Nothing has been changed yet, so there is nothing to undo."; }

    const target = resolveInside(root, last.path);
    const now = fs.existsSync(target) ? fs.readFileSync(target, "utf8") : null;
    if (now !== null && now !== last.after) {
        history.push(last);
        return last.path + " has changed since the agent wrote it, so it was left "
            + "alone rather than overwriting whatever that was.";
    }

    if (!last.existed) {
        try { fs.unlinkSync(target); } catch { /* already gone */ }
        return "Removed " + last.path + ", which the agent had created.";
    }
    fs.writeFileSync(target, last.before, "utf8");
    return "Put " + last.path + " back as it was.";
}

function writeFile(root: string, args: Record<string, string>): string {
    const target = resolveInside(root, args.path);
    fs.mkdirSync(path.dirname(target), { recursive: true });
    const content = args.content ?? '';
    const existed = fs.existsSync(target);
    const before = existed ? fs.readFileSync(target, 'utf8') : '';
    fs.writeFileSync(target, content, 'utf8');
    record({ path: args.path, before, after: content, existed, at: Date.now() });

    if (!existed) {
        // A trailing newline is not a line. Counting the split parts called a
        // one-line file two lines, every time.
        const lines = content.split(String.fromCharCode(10));
        if (lines.length && lines[lines.length - 1] === '') { lines.pop(); }
        const count = lines.length;
        return "Created " + args.path + " (" + count + (count === 1 ? " line)." : " lines).");
    }
    const shown = unified(before, content, args.path);
    return shown || (args.path + " was written with exactly what was already in it.");
}

function editFile(root: string, args: Record<string, string>): string {
    const target = resolveInside(root, args.path);
    if (!fs.existsSync(target)) {
        return `${args.path} does not exist.`;
    }
    const current = fs.readFileSync(target, 'utf8');
    const find = args.find ?? '';
    // Counting with split rather than a regex: the text is code, and code is
    // full of characters a regex would read as syntax.
    const occurrences = find ? current.split(find).length - 1 : 0;
    if (occurrences === 0) {
        return `That exact text is not in ${args.path}. Read the file again and copy the lines you mean, including their indentation.`;
    }
    if (occurrences > 1) {
        return `That text appears ${occurrences} times in ${args.path}, so it is not clear which one you mean. Include more surrounding lines to make it unique.`;
    }
    const updated = current.replace(find, args.replace ?? '');
    fs.writeFileSync(target, updated, 'utf8');
    record({ path: args.path, before: current, after: updated, existed: true, at: Date.now() });
    const shown = unified(current, updated, args.path);
    return shown || (args.path + " is unchanged - the replacement matches what was there.");
}

function search(root: string, args: Record<string, string>): string {
    const under = args.path ? resolveInside(root, args.path) : root;
    const query = args.query ?? '';
    const hits: string[] = [];
    for (const shown of walk(root, under)) {
        if (shown.endsWith('/') || hits.length >= 80) {
            continue;
        }
        const full = path.join(root, shown);
        let text: string;
        try {
            if (fs.statSync(full).size > MAX_READ_BYTES) {
                continue;
            }
            text = fs.readFileSync(full, 'utf8');
        } catch {
            continue;
        }
        const lines = text.split('\n');
        for (let i = 0; i < lines.length; i += 1) {
            if (lines[i].includes(query)) {
                hits.push(`${shown}:${i + 1}: ${lines[i].trim().slice(0, 160)}`);
                break;
            }
        }
    }
    return hits.length ? clip(hits.join('\n')) : `No file contains ${JSON.stringify(query)}.`;
}

function runCommand(root: string, args: Record<string, string>): Promise<string> {
    return new Promise((resolve) => {
        exec(
            args.command ?? '',
            { cwd: root, timeout: 300_000, maxBuffer: 8 * 1024 * 1024, windowsHide: true },
            (error, stdout, stderr) => {
                const output = `${stdout || ''}${stderr || ''}`.trim();
                if (error && (error as NodeJS.ErrnoException & { killed?: boolean }).killed) {
                    resolve('The command was still running after five minutes and was stopped.');
                    return;
                }
                // The exit code is part of the result. A command that failed
                // and a command that printed nothing have to look different,
                // or the model cannot tell whether its test passed.
                const code = error && typeof error.code === 'number' ? error.code : 0;
                resolve(clip(`exit code ${code}\n${output || '(no output)'}`));
            },
        );
    });
}

export async function execute(
    name: string,
    args: Record<string, string>,
    root: string,
): Promise<string> {
    const tool = TOOLS[name];
    if (!tool) {
        return `There is no tool called ${JSON.stringify(name)}. The ones that exist are: ${Object.keys(TOOLS).join(', ')}.`;
    }
    const missing = tool.args.filter((a) => a !== 'path' && !(a in args));
    if (missing.length) {
        return `${name} needs ${missing.join(' and ')}.`;
    }

    try {
        switch (name) {
            case 'list_files': return listFiles(root, args);
            case 'read_file': return readFile(root, args);
            case 'write_file': return writeFile(root, args);
            case 'edit_file': return editFile(root, args);
            case 'search': return search(root, args);
            case 'run_command': return await runCommand(root, args);
            case 'git': return await runCommand(root, { command: `git ${args.subcommand ?? ''}` });
            case 'open_browser': return await browser.open(String(args.url ?? ''));
            case 'browser_check': return await browser.check();
            case 'browser_click': return await browser.click(String(args.text ?? ''));
            case 'browser_type': return await browser.type(
                String(args.into ?? ''), String(args.text ?? ''),
                /^(y|yes|true|1)$/i.test(String(args.enter ?? '')));
            case 'browser_reload': return await browser.reload();
            case 'todo': {
                /* The list is shown by the panel, from the event the loop
                   emits. What goes back to the model is the list as it was
                   understood - so a model whose formatting was not read the
                   way it meant can see that and correct it, instead of
                   ticking items in its head that nobody else can see. */
                const items = parseTodo(args.items);
                if (!items.length) {
                    return 'No steps were read from that. Put one step per line, each '
                        + 'starting with [ ], [~] or [x].';
                }
                const done = items.filter((i) => i.state === 'done').length;
                return `Step list updated - ${done} of ${items.length} done:\n`
                    + items.map((i) => `${i.state === 'done' ? '[x]'
                        : i.state === 'doing' ? '[~]' : '[ ]'} ${i.text}`).join('\n');
            }
            default: return `${name} is listed but not implemented.`;
        }
    } catch (error) {
        // The failure goes to the model as a result, not up as an exception:
        // being told the path was refused is how it corrects itself.
        return `${name} failed: ${(error as Error).message.slice(0, 300)}`;
    }
}
